runner_test.go

```diff
@@ -367,6 +357,250 @@ func (*RunnerSuite) TestFatalErrorWhileSelfStartWorker(c *gc.C) {
 	c.Assert(err, gc.Equals, fatalStarter.startErr)
 }
 
+func (*RunnerSuite) TestWorkerWithNoWorker(c *gc.C) {
+	runner := worker.NewRunner(worker.RunnerParams{
+		IsFatal:      allFatal,
+		RestartDelay: time.Millisecond,
+	})
+	w, err := runner.Worker("id", nil)
+	c.Assert(errors.Cause(err), gc.Equals, worker.ErrNotFound)
+	c.Assert(w, gc.Equals, nil)
+}
+
+func (*RunnerSuite) TestWorkerWithWorkerImmediatelyAvailable(c *gc.C) {
+	runner := worker.NewRunner(worker.RunnerParams{
+		IsFatal:      allFatal,
+		RestartDelay: time.Millisecond,
+	})
+	starter := newTestWorkerStarter()
+	runner.StartWorker("id", starter.start)
+
+	// Wait long enough to be reasonably confident that the
+	// worker has been started and registered.
+	time.Sleep(10 * time.Millisecond)
+
+	stop := make(chan struct{})
+	close(stop)
+	w, err := runner.Worker("id", stop)
+	c.Assert(err, gc.Equals, nil)
+	c.Assert(w, gc.NotNil)
+	c.Assert(w.(*testWorker).starter, gc.Equals, starter)
+}
+
+func (*RunnerSuite) TestWorkerWithWorkerNotImmediatelyAvailable(c *gc.C) {
+	runner := worker.NewRunner(worker.RunnerParams{
+		IsFatal:      allFatal,
+		RestartDelay: time.Millisecond,
+	})
+	defer worker.Stop(runner)
+	starter := newTestWorkerStarter()
+	startCh := make(chan struct{})
+	runner.StartWorker("id", func() (worker.Worker, error) {
+		startCh <- struct{}{}
+		<-startCh
+		return starter.start()
+	})
+	<-startCh
+	// The start function has been called but we know the
+	// worker hasn't been returned yet.
+	wc := make(chan worker.Worker)
+	go func() {
+		w, err := runner.Worker("id", nil)
+		c.Check(err, gc.Equals, nil)
+		wc <- w
+	}()
+	// Sleep long enough that we're pretty sure that Worker
+	// will be blocked.
+	time.Sleep(10 * time.Millisecond)
```

Comment:
Same deal - there's nothing externally visible that tells us whether Worker has blocked or not.